### ISSUE TITLE
feat: add tqdm progress bars to iterative methods

### ISSUE THREAD

**Issue by jejjohnson** [jejjohnson/rbig#50]
## Problem

RBIG is an **iterative** algorithm — fitting involves looping through potentially hundreds of layers, each performing marginal Gaussianization + rotation. With `n_layers=100` (the default) or higher, `fit()` can take minutes on large datasets with **zero user feedback**. The same applies to `transform()`, `inverse_transform()`, `score_samples()`, and `jacobian()` which loop over all fitted layers at inference time.

Currently there is **no progress indication** anywhere in the codebase. Users staring at a silent terminal have no idea whether the model is 5% or 95% done, whether early stopping kicked in, or how fast each layer is processing.

### Affected Classes and Methods

| Class | Method | Loop | Iterations | Notes |
|-------|--------|------|------------|-------|
| **`AnnealedRBIG`** | `fit()` | `for i in range(self.n_layers)` | Up to 100 (default) | **Primary target** — slowest, has early stopping |
| **`AnnealedRBIG`** | `transform()` | `for layer in self.layers_` | `len(self.layers_)` | Forward pass through all fitted layers |
| **`AnnealedRBIG`** | `inverse_transform()` | `for layer in reversed(self.layers_)` | `len(self.layers_)` | Reverse pass |
| **`AnnealedRBIG`** | `score_samples()` | `for layer in self.layers_` | `len(self.layers_)` | Forward pass + log-det accumulation |
| **`AnnealedRBIG`** | `jacobian()` | Two loops: layers + seed dims | `len(self.layers_)` + `n_features` | Heaviest inference method |
| **`ImageRBIG`** | `fit()` | `for i in range(self.n_layers)` | Up to `n_layers` | No early stopping |
| **`ImageRBIG`** | `transform()` | `for marginal, rotation in self.layers_` | `len(self.layers_)` | Forward pass |
| **`ImageRBIG`** | `inverse_transform()` | `for ... in reversed(self.layers_)` | `len(self.layers_)` | Reverse pass |
| **`XarrayRBIG`** | `fit()` / `transform()` / etc. | Delegates to `AnnealedRBIG` | — | Inherits progress from underlying model |

---

### Constructor Parameter

Add a `verbose` parameter to `AnnealedRBIG` and `ImageRBIG`:

```python
class AnnealedRBIG:
    def __init__(
        self,
        n_layers: int = 100,
        rotation: str = "pca",
        zero_tolerance: int = 60,
        tol: float | str = 1e-5,
        random_state: int | None = None,
        strategy: list | None = None,
        verbose: bool | int = False,       # NEW
    ):
        ...
        self.verbose = verbose
```

**Semantics:**
- `verbose=False` or `verbose=0` — no progress bars (default, backward compatible)
- `verbose=True` or `verbose=1` — show progress bars for `fit()` only
- `verbose=2` — show progress bars for `fit()`, `transform()`, `inverse_transform()`, `score_samples()`, `jacobian()`

Rationale: `fit()` is the slow one users always want feedback on. Inference methods are usually faster and showing bars on every `transform()` call can be noisy, so `verbose=2` is opt-in.

### Usage Examples

```python
from rbig import AnnealedRBIG

# Silent (default, no change)
model = AnnealedRBIG(n_layers=100)

# Show fit() progress
model = AnnealedRBIG(n_layers=100, verbose=True)

# Show progress on everything
model = AnnealedRBIG(n_layers=100, verbose=2)
```

#### `fit()` — the main bar

```
Fitting RBIG:  45%|████▌     | 45/100 [00:12<00:14, 3.75 layers/s, TC=0.0032, δTC=2.1e-04]
```

- **desc**: `"Fitting RBIG"` / `"Fitting ImageRBIG"`
- **total**: `self.n_layers`
- **postfix**: `TC` (current total correlation), `δTC` (improvement this layer)
- **Early stopping**: when the loop breaks early, call `pbar.total = i + 1; pbar.refresh()` so the bar shows 100% at the actual stop point rather than a partial bar

#### `transform()` / `inverse_transform()` / `score_samples()`

```
Transforming:  72%|███████▏  | 36/50 [00:02<00:01, 18.0 layers/s]
```

- **desc**: `"Transforming"` / `"Inverse transforming"` / `"Scoring"`
- **total**: `len(self.layers_)`

#### `jacobian()` — two-phase bar

```
Jacobian (layers):  100%|██████████| 50/50 [00:05<00:00]
Jacobian (dims):     60%|██████    | 3/5 [00:03<00:02]
```

Or a single bar for the dominant loop (seed-dimension loop is O(n_features × n_layers) per iteration, so it's the heavier one).

---

### Step 1: Add `tqdm` as a dependency

Add to `pyproject.toml` under `[project.dependencies]`:

```toml
dependencies = [
    ...,
    "tqdm>=4.60",
]
```

`tqdm` is lightweight (~100KB), pure Python, zero transitive dependencies, and already a de-facto standard in the scientific Python ecosystem.

### Step 2: Create a helper for conditional progress bars

In `rbig/_src/_progress.py`:

```python
"""Progress bar utilities."""

from __future__ import annotations

from typing import Iterable, Iterator

from tqdm.auto import tqdm


def maybe_tqdm(
    iterable: Iterable,
    *,
    verbose: bool | int,
    level: int = 1,
    **kwargs,
) -> Iterator:
    """Wrap an iterable with tqdm if verbose >= level, else pass through.

    Parameters
    ----------
    iterable : Iterable
        The iterable to optionally wrap.
    verbose : bool or int
        Verbosity level from the model.
    level : int
        Minimum verbosity required to show this bar.
        1 = fit-only bars, 2 = all bars.
    **kwargs
        Forwarded to ``tqdm.auto.tqdm``.
    """
    v = int(verbose) if isinstance(verbose, bool) else verbose
    if v >= level:
        return tqdm(iterable, **kwargs)
    return iter(iterable)
```

Using `tqdm.auto` automatically picks the right frontend (notebook widget vs terminal).

### Step 3: Add `verbose` parameter to `AnnealedRBIG`

In `rbig/_src/model.py`:

1. Add `verbose: bool | int = False` to `__init__` signature
2. Store as `self.verbose = verbose`

### Step 4: Wrap the `fit()` loop

```python
from rbig._src._progress import maybe_tqdm

# In AnnealedRBIG.fit():
pbar = maybe_tqdm(
    range(self.n_layers),
    verbose=self.verbose,
    level=1,
    desc="Fitting RBIG",
    total=self.n_layers,
)
for i in pbar:
    ...
    # After computing TC:
    if hasattr(pbar, 'set_postfix'):
        postfix = {"TC": f"{tc:.4g}"}
        if i > 0:
            postfix["δTC"] = f"{delta:.2e}"
        pbar.set_postfix(postfix)

    # On early stopping:
    if zero_count >= self.zero_tolerance:
        if hasattr(pbar, 'total'):
            pbar.total = i + 1
            pbar.refresh()
        break
```

### Step 5: Wrap inference loops (verbose >= 2)

For `transform()`, `inverse_transform()`, `score_samples()`:

```python
layers = maybe_tqdm(
    self.layers_,
    verbose=self.verbose,
    level=2,
    desc="Transforming",
    total=len(self.layers_),
)
for layer in layers:
    ...
```

### Step 6: Wrap `jacobian()` loops

The seed-dimension loop is the outer loop. Wrap it with verbose >= 2:

```python
dims = maybe_tqdm(
    range(n_features),
    verbose=self.verbose,
    level=2,
    desc="Jacobian",
    total=n_features,
)
for idim in dims:
    ...
```

### Step 7: Add `verbose` to `ImageRBIG`

Same pattern — add `verbose` parameter, wrap `fit()` / `transform()` / `inverse_transform()`.

### Step 8: Thread `verbose` through `XarrayRBIG`

`XarrayRBIG` delegates to `AnnealedRBIG`. Add `verbose` to its `__init__` and pass it through:

```python
class XarrayRBIG:
    def __init__(self, ..., verbose: bool | int = False):
        ...
        self.verbose = verbose

    def fit(self, X):
        ...
        kwargs["verbose"] = self.verbose  # forward to AnnealedRBIG
```

### Step 9: Tests

- Test that `verbose=False` produces no output (capture stderr)
- Test that `verbose=True` produces tqdm output during `fit()`
- Test that `verbose=2` produces output during `transform()` too
- Test that early stopping still shows a complete bar
- Test that `maybe_tqdm` returns a plain iterator when verbose=0

### Step 10: Update `__init__.py` exports (if needed)

No new public API to export — `verbose` is just a constructor parameter.

---

### 1. Notebook vs Terminal Detection

`tqdm.auto` handles this automatically — it uses `tqdm.notebook` in Jupyter and `tqdm.std` in terminals. This is important for RBIG since many users are likely working in Jupyter notebooks for data science workflows.

### 2. `fit_transform()` Double Bars

`AnnealedRBIG.fit_transform()` calls `self.fit(X).transform(X)`. With `verbose=2`, this would show a fit bar AND then a transform bar. This is actually fine and expected — they're separate operations. But worth noting.

### 3. Nested Progress Bars for `XarrayRBIG` with Multiple Variables

If `XarrayRBIG` ever processes multiple variables in a loop (e.g., fitting per-variable models), you'd want an outer bar for variables and inner bar for layers. Currently it fits a single model, so this isn't an issue today, but worth keeping in mind.

### 4. `score_samples()` Re-traversal

`score()` calls `score_samples()`, which loops through all layers. If someone calls `score()` with `verbose=2`, they'll see a progress bar. That's correct behavior.

### 5. `predict_proba()` → `jacobian()` Chain

`predict_proba()` calls `jacobian()` internally. With `verbose=2`, the Jacobian bar will appear when calling `predict_proba()`. This is desirable — `predict_proba()` with the full Jacobian is the slowest inference method.

### 6. The `log_det_jacobian` Accumulation in `fit()`

The `fit()` loop computes `layer.log_det_jacobian(Xt)` on every iteration. This per-layer computation is part of the cost the progress bar is tracking, so no separate bar is needed for it.

### 7. Thread Safety / Parallel Execution

If users ever parallelize RBIG (e.g., multiple `XarrayRBIG` fits with joblib), multiple tqdm bars can conflict on stderr. `tqdm` handles this with `position` parameter, but it's not something we need to solve now. Just don't use any global state in the progress bar helper.

### 8. `tqdm` as Required vs Optional Dependency?

**Proposed decision**: make it **required**, not optional. Rationale:
- tqdm is 100KB, pure Python, zero dependencies
- It's already installed in virtually every scientific Python environment
- Making it optional adds complexity (try/except imports, fallback paths) for negligible benefit
- sklearn, pandas, huggingface, and pytorch all use tqdm

**Open question**: Should we make it optional via an extras group (e.g., `pip install rbig[progress]`) instead? This adds complexity but keeps the core dependency list minimal.

### 9. Per-Call `verbose` Override

Some advanced users may want to toggle progress per-call rather than per-model:

```python
model.fit(X, verbose=True)  # override model-level setting
```

**Open question**: Is this worth the API complexity now, or defer to a future enhancement? Constructor-level `verbose` matches sklearn's pattern and is simpler.

### 10. Early Stopping Feedback

The `fit()` progress bar's postfix already shows `TC` and `δTC`. When early stopping triggers, it might be nice to also print a one-line message:

```python
if zero_count >= self.zero_tolerance:
    if self.verbose:
        tqdm.write(f"Early stopping at layer {i+1}/{self.n_layers} (TC={tc:.4g})")
    break
```

**Open question**: Is this too noisy, or is the explicit early-stopping message valuable?

---

## File Change Summary

| File | Action | Notes |
|------|--------|-------|
| `pyproject.toml` | Update | Add `tqdm>=4.60` to dependencies |
| `rbig/_src/_progress.py` | **New** | `maybe_tqdm()` helper |
| `rbig/_src/model.py` | Update | Add `verbose` to `AnnealedRBIG`, wrap 5 loops |
| `rbig/_src/image.py` | Update | Add `verbose` to `ImageRBIG`, wrap 3 loops |
| `rbig/_src/xarray_st.py` | Update | Add `verbose` to `XarrayRBIG`, forward to underlying model |
| `tests/test_progress.py` | **New** | Tests for progress bar behavior |

**Total**: 2 new files, 4 updated files

---

**Plan reference**: `rbig/plans/progress_bars.md`
